### PR TITLE
feat(monitors): Add intervals and various UI

### PR DESCRIPTION
**Pull request by dcramer**
- Add interval schedule types
- Add monitor update endpoint
- Add monitor delete endpoint
- Add monitor create endpoint
- Add pause/enable action to monitor header (ui)
- Add delete action to monitor header (ui)
- Expose config on monitor edit (ui) - cron job only
- Add create monitor flow
- Support non-standard formats (converted on input)

### CODE REVIEW

**Comment by dcramer** [comment]
screen-shot-2019-02-09-14-05-50 89

**Comment by dcramer** [comment]
This was a little painful to map things to the child config attribute, both in the UI and backend.  I hadn't used this model behavior before (in `<Form>`), but this seemed like the cleanest approach. UI only supports cron job and crontab format right now.

**Comment by dcramer** [comment]
There's some issues that need sorted yet:

1. the form Model doesn't take on initial values (e.g. with SelectField grabbing the first value as a default)
2. (not our concern right now) onSubmitSuccess and likely other callbacks don't propagate errors upwards
3. We need Access.has_project_scope - can't be asked to write another one-off project permissions check

File: src/sentry/api/validators/monitor.py
Comment (dcramer) [inline src/sentry/api/validators/monitor.py]: also this syntax would have the project passed into it, so its still wrong

File: src/sentry/static/sentry/app/views/organizationMonitors/monitorForm.jsx
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationMonitors/monitorForm.jsx]: You don't need to do this, passing `initialData` to `<Form>` is sufficient

File: src/sentry/static/sentry/app/views/organizationMonitors/monitorForm.jsx
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationMonitors/monitorForm.jsx]: So this probably won't work unless you wrap this block with an `<Observer>` (from `mobx-react` package).

```
<Observer>
{() => {
return this.form.getValue('type') === 'cron_job' &&
                       <SelectField
                        name="config.schedule_type"
                        label={t('Schedule Type')}
                        disabled={!hasAccess}
                        choices={[['crontab', 'Crontab']]}
                        required
                      />
}}
</Observer>
```

(You could wrap the entire form, but that means the entire form re-renders when the `type` field changes, instead of just this field)

File: src/sentry/static/sentry/app/views/organizationMonitors/monitorHeaderActions.jsx
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationMonitors/monitorHeaderActions.jsx]: If you use `withApi` HoC on `MonitorHeaderActions`, you'll get instance of `Client` in props and it'll operate the same as the `ApiMixin` where it cancels requests when we unmount the component.

File: src/sentry/static/sentry/app/views/organizationMonitors/monitorModel.jsx
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationMonitors/monitorModel.jsx]: Don't you need to set data for keys that don't start with `config`?

File: src/sentry/static/sentry/app/views/organizationMonitors/monitorModel.jsx
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationMonitors/monitorModel.jsx]: Hmmm I don't think you need this, this gets called by `FormModel.getTransformedData` which you've overridden

Comment (dcramer) [reply src/sentry/static/sentry/app/views/organizationMonitors/monitorModel.jsx]: hrm might have done this before I realized the default getTransformedData was problematic for me (was using getData originally)

File: src/sentry/static/sentry/app/views/organizationMonitors/monitorModel.jsx
Comment (dcramer) [reply src/sentry/static/sentry/app/views/organizationMonitors/monitorModel.jsx]: oops - didnt test the non-config case clearly

File: src/sentry/static/sentry/app/views/organizationMonitors/monitorForm.jsx
Comment (dcramer) [reply src/sentry/static/sentry/app/views/organizationMonitors/monitorForm.jsx]: What's Observer do exactly for the layman? Seems like thats what I need. Also are we intending to keep using mobx or should I do something thats future proof?

File: src/sentry/static/sentry/app/views/organizationMonitors/monitorForm.jsx
Comment (dcramer) [reply src/sentry/static/sentry/app/views/organizationMonitors/monitorForm.jsx]: ah does Form bind initialData on model? I only tried the inverse way.

File: src/sentry/static/sentry/app/views/organizationMonitors/monitorForm.jsx
Comment (billyvg) [reply src/sentry/static/sentry/app/views/organizationMonitors/monitorForm.jsx]: Undetermined, I would say a non-mobx specific approach would be better (e.g. onChange and set react state).

File: src/sentry/static/sentry/app/views/organizationMonitors/monitorModel.jsx
Comment (dcramer) [reply src/sentry/static/sentry/app/views/organizationMonitors/monitorModel.jsx]: it gets called in saveFieldRequest as well

Comment (billyvg) [reply src/sentry/static/sentry/app/views/organizationMonitors/monitorModel.jsx]: that's if you use save on blur functionality

File: src/sentry/static/sentry/app/views/organizationMonitors/monitorForm.jsx
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/organizationMonitors/monitorForm.jsx]: > What's Observer do exactly for the layman

Mobx `Observer` will watch for observable values to be accessed within the function, it keeps track of which ones get accessed and will trigger a re-render when the observable is changed. In this case, using `form.getValue()` looks up the observable fields value.

https://github.com/getsentry/sentry/blob/cc608632ee5f0921954502f3098f9a8c9d39b42c/src/sentry/static/sentry/app/views/settings/components/forms/model.jsx#L11-L14

File: src/sentry/tasks/check_monitors.py
Comment (markstory) [inline src/sentry/tasks/check_monitors.py]: Is the checkin timeout something that users will want to tweak?

Comment (dcramer) [reply src/sentry/tasks/check_monitors.py]: probably, but the intention is to not expose every possibility at this stage

File: src/sentry/static/sentry/app/views/organizationMonitors/monitorForm.jsx
Comment (dcramer) [reply src/sentry/static/sentry/app/views/organizationMonitors/monitorForm.jsx]: yes, but likely as a followup PR

File: src/sentry/static/sentry/app/views/organizationMonitors/monitorModel.jsx
Comment (dcramer) [reply src/sentry/static/sentry/app/views/organizationMonitors/monitorModel.jsx]: Ill keep it for correctness then

File: src/sentry/static/sentry/app/views/organizationMonitors/monitorForm.jsx
Comment (dcramer) [reply src/sentry/static/sentry/app/views/organizationMonitors/monitorForm.jsx]: k im just gonna use Observer because i dont want to spend much time on this

**Comment by dcramer** [comment]
jk we sitll have the issues w/ the access check

**Comment by dcramer** [comment]
Ok so the only thing I'm aware of that's left is getting project access scoped. I might hack on that since otherwise I dont know when this will land. Can probably put in something very basic that replicates the other code for now, and leave it up to others to refactor/expand funtionality.

**Comment by dcramer** [comment]
Blocked on #12063

**Comment by dcramer** [comment]
Good to go. Project selection works as expected.

**Review by markstory — APPROVED** [review]
🍪

**Pull request merged**
